frontend/src/components/Cart.tsx

```diff
@@ -86,123 +85,165 @@ const ContentStyle = {
-    refetch: refetchCoupon,
+    refetch: refetchUsableCoupon,
   } = useQuery({
-    queryKey: ['usableCouponData'],
+    queryKey: ['BuyerGetUsableCoupon', cartInfo.id],
     queryFn: async () => {
-      const response = await fetch('/resources/UsableCoupons.json');
+      const response = await fetch(`/api/buyer/cart/${cartInfo.id}/coupon`, {
+        method: 'GET',
+        headers: {
+          Authorization: `Bearer ${token}`,
+          Accept: 'application/json',
+        },
+      });
       RouteOnNotOK(response, navigate);
-      return response.json();
+      return (await response.json()) as IUsableCoupon[];
     },
-    select: (data) => data as UsableCouponProps[],
     refetchOnWindowFocus: false,
     enabled: false,
   });
-  const onRefetchUsableCoupon = () => {
-    console.log('refetch usable coupon');
-    refetchCoupon();
-    if (usableCouponStatus === 'pending') {
-      return <div>Loading...</div>;
-    }
-    if (usableCouponStatus === 'error') {
-      return <div>Error fetching data</div>;
-    }
-  };
 
   const onViewCheckout = () => {
-    // TODO: Buyer get checkout
-    // GET /buyer/cart/:cart_id/checkout
-    onRefetchCheckout();
+    // check if the product is still available
+    for (let i = 0; i < products.length; i++) {
+      if (products[i].stock === 0) {
+        alert(`${products[i].name} is out of stock, please remove it from the cart`);
+        return;
+      }
+      if (products[i].stock < products[i].quantity) {
+        alert(
+          `Stock for ${products[i].name} is insufficient, please reduce the quantity. ( can only supply ${products[i].stock})`,
+        );
+        return;
+      }
+      if (!products[i].enabled) {
+        alert(`${products[i].name} is disabled, please remove it from the cart`);
+        return;
+      }
+    }
+    refetchCheckout();
     setCanvaShow(true);
   };
-  const onPay = () => {
-    // TODO: Buyer checkout
-    // POST /buyer/cart/:cart_id/checkout
+
+  const onChooseCoupon = () => {
+    refetchUsableCoupon();
+    setModalShow(true);
+  };
+
+  const onPay = async () => {
     if (watch('card_id') === null) {
-      // TODO: implement real card selection
-      console.log('please select a card');
+      if (checkoutData?.payments.length === 0) {
+        alert('please add a card');
+      } else {
+        alert('please select a card');
+      }
       return;
     }
-    console.log('pay');
-    onRefetch();
-    setCanvaShow(false);
-  };
-  const onChooseCoupon = () => {
-    // TODO: Buyer get usable coupon of cart
-    // GET /buyer/cart/:cart_id/coupon
-    onRefetchUsableCoupon();
-    setModalShow(true);
+    const resp = await fetch(`/api/buyer/cart/${cartInfo.id}/checkout`, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${token}`,
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(checkoutData?.payments[watch('card_id')]),
```

Comment:
is POST null JSON expected behavior?

Reply:
no, and i think ill add a comment explan this

Reply:
> no

then I think you should check `if checkoutData === null or undefined` before the POST.
if adding `?` everywhere is a JS convention, then I have no objections.

Reply:
yes im doing that rn
but not the element part tho